src/KafkaFlow.IntegrationTests/OpenTelemetryTests.cs

```diff
@@ -0,0 +1,174 @@
+    using Microsoft.Extensions.Hosting;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class OpenTelemetryTests
+    {
+        private readonly Fixture fixture = new();
+
+        private List<Activity> exportedItems;
+
+        private IServiceProvider provider;
+
+        [TestInitialize]
+        public void Setup()
+        {
+            this.exportedItems = new List<Activity>();
+            this.provider = this.GetServiceProvider();
+            MessageStorage.Clear();
+        }
+
+        [TestMethod]
+        public async Task AddOpenTelemetry_ProducingAndConsumingOneMessage_TraceAndSpansAreCreatedCorrectly()
+        {
+            // Arrange
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddSource("KafkaFlow.OpenTelemetry")
+            .AddInMemoryExporter(this.exportedItems)
+            .Build();
+
+            var producer = this.provider.GetRequiredService<IMessageProducer<GzipProducer>>();
+            var message = this.fixture.Create<byte[]>();
+
+            // Act
+            await producer.ProduceAsync(null, message);
+
+            // Assert
+            await Task.Delay(8000).ConfigureAwait(false);
+
+            Assert.IsNotNull(this.exportedItems);
+
+            var producerSpan = this.exportedItems.Find(x => x.Kind == ActivityKind.Producer);
+            var consumerSpan = this.exportedItems.Find(x => x.Kind == ActivityKind.Consumer);
+
+            Assert.IsNull(producerSpan.ParentId);
+            Assert.AreEqual(producerSpan.TraceId, consumerSpan.TraceId);
+            Assert.AreEqual(consumerSpan.ParentSpanId, producerSpan.SpanId);
+        }
+
+        [TestMethod]
+        public async Task AddOpenTelemetry_ProducingAndConsumingOneMessage_BaggageIsPropagatedFromTestActivityToConsumer()
+        {
+            // Arrange
+            var kafkaFlowTestString = "KafkaFlowTest";
+            var baggageName1 = "TestBaggage1";
+            var baggageValue1 = "TestBaggageValue1";
+            var baggageName2 = "TestBaggage2";
+            var baggageValue2 = "TestBaggageValue2";
+
+            using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddSource("KafkaFlow.OpenTelemetry")
+            .AddSource(kafkaFlowTestString)
+            .AddInMemoryExporter(this.exportedItems)
+            .Build();
+
+            var producer = this.provider.GetRequiredService<IMessageProducer<GzipProducer>>();
+            var message = this.fixture.Create<byte[]>();
+
+            // Act
+            ActivitySource activitySource = new(kafkaFlowTestString);
+
+            var activity = activitySource.StartActivity("TestActivity", ActivityKind.Client);
+
+            activity.AddBaggage(baggageName1, baggageValue1);
+            activity.AddBaggage(baggageName2, baggageValue2);
+
+            await producer.ProduceAsync(null, message);
+
+            // Assert
+            await Task.Delay(40000).ConfigureAwait(false);
```

Comment:
Is there any reasoning for this high delay?

Reply:
Please replace sleep's with retry policies with timeout.

Reply:
Same as https://github.com/Farfetch/kafkaflow/pull/451#discussion_r1360801150.

Reply:
lgtm